crates/noirc_frontend/src/parser/parser.rs

```diff
@@ -625,6 +636,58 @@ where
         })
 }
 
+/// [a; N]
+fn array_sugar<P>(expr_parser: P) -> impl NoirParser<ExpressionKind>
+where
+    P: ExprParser,
+{
+    expr_parser
+        .then_ignore(just(Token::Semicolon))
+        .then(literal())
+        .delimited_by(just(Token::LeftBracket), just(Token::RightBracket))
+        .validate(|(lhs, count), span, emit| {
+            let count = validate_array_count(count, span, emit);
+
+            // Desugar the array by replicating the lhs 'count' times. TODO: This is inefficient
```

Comment:
I think what we could do if you want to make it more efficient is to keep the compact form `[expr; count]` up to the backend, which will generate `count` store operations anyway so I am not sure the gain will be that relevant.